backend/postgres/postgres.go

```diff
@@ -0,0 +1,955 @@
+			return nil, ctx.Err()
+		}
+		if err == sql.ErrNoRows {
+			// No new activity tasks to process
+			return nil, backend.ErrNoWorkItems
+		}
+		return nil, fmt.Errorf("failed to scan the activity work-item: %w", err)
+	}
+
+	e := new(protos.HistoryEvent)
+	err := protojson.Unmarshal(eventPayload, e)
+	if err != nil {
+		return nil, err
+	}
+
+	wi := &backend.ActivityWorkItem{
+		SequenceNumber: sequenceNumber,
+		InstanceID:     api.InstanceID(instanceID),
+		NewEvent:       e,
+		LockedBy:       be.workerName,
+	}
+	return wi, nil
+}
+
+func (be *postgresBackend) CompleteActivityWorkItem(ctx context.Context, wi *backend.ActivityWorkItem) error {
+	if err := be.ensureDB(ctx); err != nil {
+		return err
+	}
+
+	tx, err := be.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	bytes, err := protojson.Marshal(wi.Result)
+	if err != nil {
+		return err
+	}
+
+	_, err = tx.ExecContext(ctx, "INSERT INTO durabletask.NewEvents (InstanceID, EventPayload) VALUES ($1, $2)", string(wi.InstanceID), bytes)
+	if err != nil {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		return fmt.Errorf("failed to insert into NewEvents table: %w", err)
+	}
+
+	dbResult, err := tx.ExecContext(ctx, "DELETE FROM durabletask.NewTasks WHERE SequenceNumber = $1 AND LockedBy = $2", wi.SequenceNumber, wi.LockedBy)
+	if err != nil {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		return fmt.Errorf("failed to delete from NewTasks table: %w", err)
+	}
+
+	rowsAffected, err := dbResult.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("failed get rows affected by delete statement: %w", err)
+	} else if rowsAffected == 0 {
+		return backend.ErrWorkItemLockLost
+	}
+
+	if err = tx.Commit(); err != nil {
+		return fmt.Errorf("failed to commit transaction: %w", err)
+	}
+
+	return nil
+}
+
+func (be *postgresBackend) AbandonActivityWorkItem(ctx context.Context, wi *backend.ActivityWorkItem) error {
+	if err := be.ensureDB(ctx); err != nil {
+		return err
+	}
+
+	dbResult, err := be.db.ExecContext(
+		ctx,
+		"UPDATE durabletask.NewTasks SET LockedBy = NULL, LockExpiration = NULL WHERE SequenceNumber = $1 AND LockedBy = $2",
+		wi.SequenceNumber,
+		wi.LockedBy,
+	)
+	if err != nil {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		return fmt.Errorf("failed to update the NewTasks table for abandon: %w", err)
+	}
+
+	rowsAffected, err := dbResult.RowsAffected()
+	if err != nil {
+		return fmt.Errorf("failed get rows affected by update statement for abandon: %w", err)
+	} else if rowsAffected == 0 {
+		return backend.ErrWorkItemLockLost
+	}
+
+	return nil
+}
+
+func (be *postgresBackend) PurgeOrchestrationState(ctx context.Context, id api.InstanceID) error {
+	if err := be.ensureDB(ctx); err != nil {
+		return err
+	}
+
+	tx, err := be.db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	row := tx.QueryRowContext(ctx, "SELECT 1 FROM durabletask.Instances WHERE InstanceID = $1", string(id))
```

Comment:
Here's another option to write this method without using transactions and a single atomic query:

1. For the History table, make InstanceID a foreign key reference to Instances.InstanceID, with an "ON DELETE CASCADE". This means that if the row is deleted from Instances, it's automatically (and atomically) deleted from History too.
2. Next, use this query that performs both a lookup and a deletion atomically, without the need for a transaction:


```sql
WITH del AS (
	DELETE
		FROM durabletask.Instances
		WHERE InstanceID = $1
		AND RuntimeStatus IN ('COMPLETED', 'FAILED', 'TERMINATED')
		RETURNING InstanceID
), delcount AS (
	SELECT count(*) AS count FROM del
)
SELECT delcount.count
	FROM durabletask.Instances, delcount
	WHERE durabletask.Instances.InstanceID = $1
```

This query has 3 possible states:

1. No rows returned: this means the instance doesn't exist (you can return `api.ErrInstanceNotFound`)
2. Returned value is "0" -> The instance exists but it's running (you can return `api.ErrNotCompleted`)
3. Returned value is "1" -> The instance exists and it's been deleted successfully (it was in completed/failed/terminated state)